### PR TITLE
revert: add verbose option

### PR DESCRIPTION
**Pull request by marksliva**
Always show revert substeps. Stream the directories being deleted to the CLI
when the verbose option is set.

Currently we are only streaming the directories that are being deleted on the
hub, but once we introduce a way to stream output to the hub from the agents,
we'll want to print the directories being deleted on each agent as well.

### CODE REVIEW

File: cli/commanders/steps.go [hunk 1]
```diff
@@ -50,6 +50,11 @@ var SubstepDescriptions = map[idl.Substep]substepText{
 	idl.Substep_FINALIZE_START_TARGET_CLUSTER:                     substepText{"Starting target cluster...", "Start target cluster"},
 	idl.Substep_FINALIZE_UPGRADE_STANDBY:                          substepText{"Upgrading standby master...", "Upgrade standby master"},
 	idl.Substep_FINALIZE_UPGRADE_MIRRORS:                          substepText{"Upgrading mirror segments...", "Upgrade mirror segments"},
+	idl.Substep_DELETE_PRIMARY_DATADIRS:                           substepText{"Deleting primary datadirs...", "Delete primary datadirs"},
```
Comment (bhuvnesh2703) [inline cli/commanders/steps.go:53]: Delete primary segments datadirs (to be in sync with Upgrade primary segments)

File: cli/commanders/steps.go [hunk 2]
```diff
@@ -50,6 +50,11 @@ var SubstepDescriptions = map[idl.Substep]substepText{
...
+	idl.Substep_DELETE_MASTER_DATADIR:                             substepText{"Deleting master datadir...", "Delete master datadir"},
+	idl.Substep_DELETE_NON_MASTER_STATEDIRS:                       substepText{"Deleting non-master state directories...", "Delete non-master state directories"},
```
Comment (bhuvnesh2703) [inline cli/commanders/steps.go:55]: Delete gpupgrade config / state directories?

Comment (kalensk) [inline cli/commanders/steps.go:55]: Maybe rephrase to something like "Deleting state directory on the segments"? Reading "non-master" is a little clunky.

File: cli/commanders/steps.go (hunk 1 above)
Comment (kalensk) [inline cli/commanders/steps.go:53]: Since this UI facing maybe change `datadirs` to `data directories`, and similarly below.

File: cli/commands/commands.go [hunk 3]
```diff
@@ -484,9 +502,37 @@ func revert() *cobra.Command {
...
+}
+
+func stopHubAndAgentsSubStep() error {
+	var err error
+
+	s := commanders.Substep(idl.Substep_STOP_HUB_AND_AGENTS)
+	defer s.Finish(&err)
+
+	err = stopHubAndAgents()
+	if err != nil {
```
Comment (bhuvnesh2703) [inline cli/commands/commands.go:529]: can potentially return  stopHubAndAgents()

File: cli/commands/commands.go [hunk 4]
```diff
@@ -453,25 +454,42 @@ func finalize() *cobra.Command {
 	return addHelpToCommand(cmd, FinalizeHelp)
 }
 
+type stdStream struct {
+}
```
Comment (kalensk) [inline cli/commands/commands.go:458]: I know its used only here, but I could also see it being in `utils` next to `DevNull`, so they are all in one place.

File: cli/commands/commands.go [hunk 5]
```diff
@@ -453,25 +454,42 @@ func finalize() *cobra.Command {
...
 			}
 
-			err = stopHubAndAgents()
+			err = stopHubAndAgentsSubStep()
 			if err != nil {
 				return err
 			}
 
-			err = upgrade.DeleteDirectories([]string{utils.GetStateDir()}, upgrade.StateDirectoryFiles)
+			err = deleteMasterStateDirectory()
```
Comment (kalensk) [inline cli/commands/commands.go:492]: Im curious if there is a way of inlining these functions since they aren't used anywhere else. It seems like it would make the code a lot cleaner.

File: hub/revert.go [hunk 6]
```diff
@@ -5,21 +5,32 @@ package hub
...
 import (
 	"github.com/greenplum-db/gpupgrade/idl"
+	"github.com/greenplum-db/gpupgrade/step"
 	"github.com/greenplum-db/gpupgrade/upgrade"
 )
 
 func (s *Server) Revert(_ *idl.RevertRequest, stream idl.CliToHub_RevertServer) (err error) {
+	st, err := step.Begin(s.StateDir, "revert", stream)
+	if err != nil {
+		return nil
```
Comment (bhuvnesh2703) [inline hub/revert.go:15]: why return nil when err is not nil?

File: utils/sys_utils.go [hunk 7]
```diff
@@ -133,3 +134,16 @@ func CreateDataDirectory(dataDir string) error {
 	}
 	return nil
 }
+
+// DevNull implements OutStreams by just discarding all writes.
+var DevNull = devNull{}
```
Comment (kalensk) [inline utils/sys_utils.go:139]: For future reference, I think it would make reviewing easier if this move was done in a separate commit.

File: hub/revert.go [hunk 8]
```diff
@@ -5,21 +5,32 @@ package hub
...
+			hostname := s.Config.Target.MasterHostname()
+
+			return upgrade.DeleteDirectories([]string{datadir}, upgrade.PostgresFiles, hostname, streams)
+		})
 	}
 
-	return DeleteStateDirectories(s.agentConns, s.Source.MasterHostname())
+	st.Run(idl.Substep_DELETE_NON_MASTER_STATEDIRS, func(_ step.OutStreams) error {
+		return DeleteStateDirectories(s.agentConns, s.Source.MasterHostname())
+	})
```
Comment (kalensk) [inline hub/revert.go:33]: Its a little confusing that the substep description is `Substep_DELETE_NON_MASTER_STATEDIRS` but then it calls `DeleteStateDirectories` with the `masterHostname`? Is that all correct?

File: hub/revert.go [hunk 9]
```diff
@@ -5,21 +5,32 @@ package hub
...
-		err = DeletePrimaryDataDirectories(s.agentConns, s.Config.Target)
-		if err != nil {
-			return err
-		}
-
-		err = upgrade.DeleteDirectories([]string{s.Config.Target.MasterDataDir()}, upgrade.PostgresFiles)
-		if err != nil {
-			return err
-		}
+		st.Run(idl.Substep_DELETE_PRIMARY_DATADIRS, func(_ step.OutStreams) error {
```
Comment (bhuvnesh2703) [inline hub/revert.go:19]: might need to review with the PM team about the description of these substeps and how many Steps we want to show. if you have already checked ignore..

File: step/step_test.go [hunk 10]
```diff
@@ -348,24 +348,24 @@ func (t *TestStore) Write(substep idl.Substep, status idl.Status) (err error) {
 	return t.WriteErr
 }
 
-// DevNull implements step.OutStreamsCloser as a no-op. It also tracks calls to
+// DevNullWithClose implements step.OutStreamsCloser as a no-op. It also tracks calls to
```
Comment (kalensk) [inline step/step_test.go:351]: Maybe not for this PR, but I wonder if all these OutStream implementations should go together since its easy to find and see which one to use. I think that might be nice.

File: upgrade/directories_test.go [hunk 11]
```diff
@@ -328,13 +348,20 @@ func TestDeleteDirectories(t *testing.T) {
 				t.Errorf("dataDir %s exists", dataDir)
 			}
 		}
+
+		expected := regexp.MustCompile("Deleting directory: .*/data/dbfast_mirror1/seg1\" on host \"localhost.local\"\nDeleting directory: \".*/data/dbfast_mirror2/seg2\" on host \"localhost.local")
+
+		actual := buf.String()
+		if !expected.MatchString(actual) {
+			t.Errorf("got stream output %s want %s", actual, expected)
+		}
```
Comment (kalensk) [inline upgrade/directories_test.go:357]: We should not be matching error strings.

Some of the other alternatives include:
1) Matching error types. For example, we do this in [cluster_test.go](https://github.com/greenplum-db/gpupgrade/blob/cb3561cb60bfacc9fd9b9bc23a923362c9df7666/greenplum/cluster_test.go#L160-L167) with [ErrInvalidSegments](https://github.com/greenplum-db/gpupgrade/blob/master/greenplum/cluster.go#L178-L180) and [InvalidSegmentsError](https://github.com/greenplum-db/gpupgrade/blob/master/greenplum/cluster.go#L230-L251).

2) Simply see if we get _any_ kind of error back. We do this in [commands_test.go](https://github.com/greenplum-db/gpupgrade/blob/master/cli/commands/commands_test.go#L52-L57)

3) Other ways...

File: step/step_test.go (hunk 10 above)
Comment (kalensk) [reply step/step_test.go:351]: Also for future reference doing this rename to DevNullWithClose in a separate commit might make things a bit easier for reviewing.

File: agent/delete_directories.go [hunk 12]
```diff
@@ -10,20 +10,25 @@ import (
...
+	"github.com/greenplum-db/gpupgrade/utils"
 )
 
 var deleteDirectories = upgrade.DeleteDirectories
 
 func (s *Server) DeleteStateDirectory(ctx context.Context, in *idl.DeleteStateDirectoryRequest) (*idl.DeleteStateDirectoryReply, error) {
 	gplog.Info("got a request to delete the state directory from the hub")
 
-	err := deleteDirectories([]string{s.conf.StateDir}, upgrade.StateDirectoryFiles)
+	hostname, _ := utils.System.Hostname()
```
Comment (kalensk) [inline agent/delete_directories.go:21]: I am not sure we should be ignoring these errors everywhere.

File: upgrade/directories_test.go (hunk 11 above)
Comment (marksliva) [reply upgrade/directories_test.go:357]: This is not testing an error message. This checks that the correct output was streamed to the buffer through `Stdout`.
```
		devNull := devNullSpy{
			outStream: &buf,
		}
```

File: step/step_test.go (hunk 10 above)
Comment (marksliva) [reply step/step_test.go:351]: I was thinking about that, however, since this is used only in testing, this seems like a better place for it than putting it in `sys_utils`, which has code that's used in both production code and the tests.

File: agent/delete_directories.go (hunk 12 above)
Comment (marksliva) [reply agent/delete_directories.go:21]: I'll remove this from the agents, since we know it is not currently being used anyway.

File: cli/commands/commands.go (hunk 5 above)
Comment (marksliva) [reply cli/commands/commands.go:492]: The main reason to do it this way was to utilize the `defer s.Finish(&err)` pattern. But I suppose we could inline it if that makes it easier to read.

File: hub/revert.go (hunk 8 above)
Comment (marksliva) [reply hub/revert.go:33]: Exactly. The `masterHostname` is used to exclude any agent that is running on the master host. The variable could be renamed to `excludeHostname` if that would make it easier to follow?

File: cli/commands/commands.go (hunk 5 above)
Comment (marksliva) [reply cli/commands/commands.go:492]: Deleting the state directory looks like this when inlined:
```
			hostname, err := os.Hostname()
			if err != nil {
				s.Finish(&err)
				return err
			}

			err = upgrade.DeleteDirectories([]string{utils.GetStateDir()}, upgrade.StateDirectoryFiles, hostname, &utils.StdStream{})
			if err != nil {
				s.Finish(&err)
				gplog.Error(err.Error())
				return err
			}
			s.Finish(&err)
```

File: cli/commanders/steps.go (hunk 2 above)
Comment (marksliva) [reply cli/commanders/steps.go:55]: What do you mean by config? By deleting the root level state directory, wouldn't that end up deleting all of the configs?

File: hub/revert.go (hunk 6 above)
Comment (marksliva) [reply hub/revert.go:15]: good catch.

File: cli/commanders/steps.go (hunk 2 above)
Comment (marksliva) [reply cli/commanders/steps.go]: Done.

File: cli/commands/commands.go (hunk 3 above)
Comment (marksliva) [reply cli/commands/commands.go]: not possible anymore, since this was inlined.

File: cli/commanders/steps.go (hunk 1 above)
Comment (marksliva) [reply cli/commanders/steps.go]: Done.

File: cli/commands/commands.go (hunk 4 above)
Comment (marksliva) [reply cli/commands/commands.go]: Moved it to utils.

File: hub/revert.go (hunk 8 above)
Comment (marksliva) [reply hub/revert.go]: I renamed it to `excludeHostname`.

File: cli/commanders/steps.go (hunk 1 above)
Comment (marksliva) [reply cli/commanders/steps.go]: Done.

File: upgrade/directories_test.go [hunk 13]
```diff
@@ -328,13 +348,20 @@ func TestDeleteDirectories(t *testing.T) {
 				t.Errorf("dataDir %s exists", dataDir)
 			}
 		}
+
+		expected := regexp.MustCompile("Deleting directory: .*/data/dbfast_mirror1/seg1\" on host \"localhost.local\"\nDeleting directory: \".*/data/dbfast_mirror2/seg2\" on host \"localhost.local")
```
Comment (kalensk) [inline upgrade/directories_test.go:352]: nite: I believe if you switch the outer quotes to back tics you won't have to escape the double quotes inside. It sometimes makes things much more readable.

File: upgrade/directories_test.go (hunk 11 above)
Comment (kalensk) [reply upgrade/directories_test.go:357]: Ah, my bad!

File: cli/commands/commands.go [hunk 14]
```diff
@@ -454,28 +454,48 @@ func finalize() *cobra.Command {
...
 			if err != nil {
 				return err
 			}
 
-			err = upgrade.DeleteDirectories([]string{utils.GetStateDir()}, upgrade.StateDirectoryFiles)
+			s = commanders.Substep(idl.Substep_DELETE_MASTER_STATEDIR)
+
+			hostname, err := os.Hostname()
 			if err != nil {
+				s.Finish(&err)
```
[the remark this answers is not in the document]
Comment (marksliva) [reply cli/commands/commands.go:488]: 2 of them can be combined, but we should only call `s.Finish(&err)` a single time in our flow. So this first `s.Finish` should stay since we return right afterwards, and the 2 after this can be combined and called first, since we'll either have an error or nil.
```
			err = upgrade.DeleteDirectories([]string{utils.GetStateDir()}, upgrade.StateDirectoryFiles, hostname, &utils.StdStream{})
			s.Finish(&err)
```

File: agent/delete_directories.go [hunk 15]
```diff
@@ -10,20 +10,21 @@ import (
...
+	"github.com/greenplum-db/gpupgrade/utils"
 )
 
 var deleteDirectories = upgrade.DeleteDirectories
 
 func (s *Server) DeleteStateDirectory(ctx context.Context, in *idl.DeleteStateDirectoryRequest) (*idl.DeleteStateDirectoryReply, error) {
 	gplog.Info("got a request to delete the state directory from the hub")
 
-	err := deleteDirectories([]string{s.conf.StateDir}, upgrade.StateDirectoryFiles)
+	err := deleteDirectories([]string{s.conf.StateDir}, upgrade.StateDirectoryFiles, "", utils.DevNull)
```
[the remark this answers is not in the document]
Comment (marksliva) [reply agent/delete_directories.go:21]: Done.

File: upgrade/directories_test.go (hunk 13 above)
Comment (marksliva) [reply upgrade/directories_test.go:352]: Done.

File: cli/commands/commands.go [hunk 16]
```diff
@@ -454,28 +454,48 @@ func finalize() *cobra.Command {
...
-			err := commanders.Revert(client)
+
+			err := commanders.Revert(client, verbose)
 			if err != nil {
 				gplog.Error(err.Error())
 				return err
 			}
 
+			s := commanders.Substep(idl.Substep_STOP_HUB_AND_AGENTS)
+
```
[the remark this answers is not in the document]
Comment (marksliva) [reply cli/commands/commands.go:476]: Done.
